src/py/flwr/server/workflow/secure_aggregation/secaggplus_workflow.py

```diff
@@ -0,0 +1,598 @@
+# Copyright 2024 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Workflow for the SecAgg+ protocol."""
+
+
+from __future__ import annotations
+
+import random
+from dataclasses import dataclass, field
+from logging import ERROR, WARN
+from typing import List, cast
+
+import flwr.common.recordset_compat as compat
+from flwr.common import (
+    Code,
+    ConfigsRecord,
+    Context,
+    FitRes,
+    Message,
+    MessageType,
+    NDArrays,
+    RecordSet,
+    Status,
+    bytes_to_ndarray,
+    log,
+    ndarrays_to_parameters,
+)
+from flwr.common.secure_aggregation.crypto.shamir import combine_shares
+from flwr.common.secure_aggregation.crypto.symmetric_encryption import (
+    bytes_to_private_key,
+    bytes_to_public_key,
+    generate_shared_key,
+)
+from flwr.common.secure_aggregation.ndarrays_arithmetic import (
+    factor_extract,
+    get_parameters_shape,
+    parameters_addition,
+    parameters_mod,
+    parameters_subtraction,
+)
+from flwr.common.secure_aggregation.quantization import dequantize
+from flwr.common.secure_aggregation.secaggplus_constants import (
+    RECORD_KEY_CONFIGS,
+    Key,
+    Stage,
+)
+from flwr.common.secure_aggregation.secaggplus_utils import pseudo_rand_gen
+from flwr.server.compat.driver_client_proxy import DriverClientProxy
+from flwr.server.compat.legacy_context import LegacyContext
+from flwr.server.driver import Driver
+
+from ..default_constant import MAIN_CONFIGS_RECORD, MAIN_PARAMS_RECORD
+from ..default_constant import Key as DefaultKey
+
+
+@dataclass
+class WorkflowState:  # pylint: disable=R0902
+    """The state of the SecAgg+ protocol."""
+
+    nid_to_fitins: dict[int, RecordSet] = field(default_factory=dict)
+    sampled_node_ids: set[int] = field(default_factory=set)
+    active_node_ids: set[int] = field(default_factory=set)
+    num_shares: int = 0
+    threshold: int = 0
+    clipping_range: float = 0.0
+    quantization_range: int = 0
+    mod_range: int = 0
+    max_weight: float = 0.0
+    nid_to_neighbours: dict[int, set[int]] = field(default_factory=dict)
+    nid_to_publickeys: dict[int, list[bytes]] = field(default_factory=dict)
+    forward_srcs: dict[int, list[int]] = field(default_factory=dict)
+    forward_ciphertexts: dict[int, list[bytes]] = field(default_factory=dict)
```

Comment:
Is there a reason for using `dict` vs `Dict` here? Asking b/c the rest of the codebase uses `Dict`

Reply:
No specific reasons... I've updated all type hints.